rcl/include/rcl/service.h

```diff
@@ -0,0 +1,319 @@
+ *
+ * If allocation is required when taking the request, e.g. if space needs to
+ * be allocated for a dynamically sized array in the target message, then the
+ * allocator given in the service options is used.
+ *
+ * The request_header is a type-erased pointer to pre-allocated a rmw struct containing
+ * meta-information about the request (e.g. the sequence number).
+ *
+ * \param[in] service the handle to the service from which to take
+ * \param[inout] request_header type-erased ptr to a request header
+ * \param[inout] ros_request type-erased ptr to an allocated ROS request message
+ * \return RCL_RET_OK if the request was taken, or
+ *         RCL_RET_INVALID_ARGUMENT if any arugments are invalid, or
+ *         RCL_RET_SERVICE_INVALID if the service is invalid, or
+ *         RCL_RET_BAD_ALLOC if allocating memory failed, or
+ *         RCL_RET_ERROR if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_take_request(
+  const rcl_service_t * service,
+  void * request_header,
+  void * ros_request);
+
+/// Send a ROS response to a client using a service.
+/* It is the job of the caller to ensure that the type of the ros_response
+ * parameter and the type associate with the service (via the type support)
+ * match.
+ * Passing a different type to send_response produces undefined behavior and cannot
+ * be checked by this function and therefore no deliberate error will occur.
+ *
+ * send_response is an non-blocking call. 
+ *
+ * The ROS response message given by the ros_response void pointer is always owned by the
+ * calling code, but should remain constant during send_response.
+ *
+ * TODO: Is this still true?
+ * This function is thread safe so long as access to both the service and the
+ * ros_response is synchronized.
+ * That means that calling rcl_send_response from multiple threads is allowed, but
+ * calling rcl_send_response at the same time as non-thread safe service functions
+ * is not, e.g. calling rcl_send_response and rcl_service_fini concurrently
+ * is not allowed.
+ * Before calling rcl_send_response the message can change and after calling
+ * rcl_send_response the message can change, but it cannot be changed during the
+ * send_response call.
+ * The same ros_response, however, can be passed to multiple calls of
+ * rcl_send_response simultaneously, even if the services differ.
+ * The ros_response is unmodified by rcl_send_response.
+ *
+ * \param[in] service handle to the service which will make the response
+ * \param[in] ros_response type-erased pointer to the ROS response message
+ * \return RCL_RET_OK if the response was sent successfully, or
+ *         RCL_RET_INVALID_ARGUMENT if any arguments are invalid, or
+ *         RCL_RET_CLIENT_INVALID if the service is invalid, or
+ *         RCL_RET_ERROR if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_send_response(
+  const rcl_service_t * service,
+  void * request_header,
```

Comment:
After some poking around I confirmed that this should indeed be called response_header and not request_header. As a side note, this is kind of annoying because the struct definition is called rmw_request_id_t, but the type is used for both response and request metadata.